redis/runtime/src/main/java/io/quarkiverse/langchain4j/redis/RedisEmbeddingStore.java

```diff
@@ -0,0 +1,246 @@
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import org.jboss.logging.Logger;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+
+import dev.langchain4j.data.document.Metadata;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.internal.Json;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import io.quarkiverse.langchain4j.QuarkusJsonCodecFactory;
+import io.quarkiverse.langchain4j.redis.runtime.RedisSchema;
+import io.quarkus.redis.datasource.ReactiveRedisDataSource;
+import io.quarkus.redis.datasource.json.ReactiveTransactionalJsonCommands;
+import io.smallrye.mutiny.Uni;
+import io.vertx.mutiny.redis.client.Command;
+import io.vertx.mutiny.redis.client.Request;
+import io.vertx.mutiny.redis.client.Response;
+
+public class RedisEmbeddingStore implements EmbeddingStore<TextSegment> {
+
+    public static final String EXTRA_ATTRIBUTES = "extra_attributes";
+    public static final String ID = "id";
+    private final ReactiveRedisDataSource ds;
+    private final RedisSchema schema;
+    private final Logger LOG = Logger.getLogger(RedisEmbeddingStore.class);
+
+    private static final String SCORE_FIELD_NAME = "vector_score";
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public RedisEmbeddingStore(ReactiveRedisDataSource ds, RedisSchema schema) {
+        this.ds = ds;
+        this.schema = schema;
+        try {
+            createIndexIfDoesNotExist();
+        } catch (Exception e) {
+            LOG.error(e);
+        }
+    }
+
+    private void createIndexIfDoesNotExist() {
+        List<String> indexes = ds.search().ft_list().await().indefinitely();
+        if (!indexes.contains(schema.getIndexName())) {
+            // TODO: rewrite to use the typesafe data source API
+            Request request = Request.cmd(Command.FT_CREATE)
+                    .arg(schema.getIndexName())
+                    .arg("ON")
+                    .arg("JSON")
+                    .arg("PREFIX")
+                    .arg("1")
+                    .arg(schema.getPrefix())
+                    .arg("SCHEMA");
+            schema.defineFields(request);
+            LOG.debug(
+                    "Creating index with command: " + request.toString().replaceAll("\r\n", " "));
+            ds.getRedis().send(request).await().indefinitely();
+        } else {
+            LOG.debug("Index in Redis already exists: " + schema.getIndexName());
```

Comment:
Yeah, it's really unfortunate. In addition, the indexing is async. 
So, it's not because you have your index that it did the job yet.